### DISCUSSION TITLE
Can't use embeddables with `forceEntityConstructor`

### DISCUSSION THREAD

**Discussion by mjroeleveld** [mikro-orm/mikro-orm#4612] [Q&A]
**Describe the bug**
Combining `discriminatorColumn` with `customType` doesn't work. It hydrates the wrong version of the object.
```
@Embeddable({
  abstract: true,
  discriminatorColumn: 'type',
})
export class FileMetadata {
  @Enum({customType: new ValueObjectStringType(FileType)})
  public readonly type: FileType;
}
```
**Versions**

| Dependency | Version |
| - | - |
| node | 16 |
| typescript | ES15 |
| mikro-orm | 5.7.4 |

**Comment by B4nan** [comment]
How does your custom type implementation look like? What do you mean by "It hydrates the wrong version of the object."? Please provide a complete repro.

Also, why are you testing on such an old ORM version? 

---

It feels a bit weird to use `@Enum` with a custom type (it feels a bit weird to use a custom type on the discriminator column as well tbh, but maybe I am missing something - which I could understand from a complete repro). Have you tried a regular `@Property` mapping?

**Reply by B4nan** [reply]
Maybe try the latest dev version too, I've been fixing one problem with enums that override the `columnType` few days ago, maybe its about that? Hard to guess what "doesn't work" actually means :]

**Reply by mjroeleveld** [reply]
Isn't 5.7 the latest version? :)

**Reply by B4nan** [reply]
5.7.4 is a pretty old version, 4 months old to be precise, which means probably more than a hundred bug fixes you are missing.

**Reply by mjroeleveld** [reply]
Ah yes, I updated however the problem persists.

**Comment by mjroeleveld** [comment]
I was slightly misled. The problem seems to be that I get back an object of type `FileMetadata`, even though the data properties of `PdfFileMetadata` are set. Meaning that I can't call method `foo()` on the returned object. Note that I use `forceEntityConstructor`.

```typescript
export enum FileTypeValue {
  Image = 'image',
  Pdf = 'pdf',
}

export class FileType {
  readonly value: string;
 
  constructor(value: string) {
    this.value = value as FileTypeValue;
  }
}

export class ValueObjectStringType extends Type<
  { toString: () => string } | undefined,
  string | undefined
> {
  constructor(private classRef: new (value: any) => { toString(): string }) {
    super();
  }

  convertToDatabaseValue(
    valueObject: { toString(): string } | string | undefined,
  ): string | undefined {
    if (valueObject === undefined) {
      return undefined;
    }
    if (typeof valueObject === 'string') {
      return valueObject;
    }
    return valueObject.toString();
  }

  convertToJSValue(value: { toString: () => string } | string | undefined):
    | {
        toString(): string;
      }
    | undefined {
    if (value === null) {
      return undefined;
    }
    return new this.classRef(
      typeof value === 'object' ? value.toString() : value,
    );
  }
}


@Embeddable({
  abstract: true,
  discriminatorColumn: 'type',
})
export abstract class FileMetadata {
  @Enum({
    customType: new ValueObjectStringType(FileType),
  })
  public readonly type: FileType;
  constructor({ type }: FileMetadataProps) {
    this.type = type;
  }
}

@Embeddable({
  discriminatorValue: FileTypeValue.Pdf,
})
export class PdfFileMetadata extends FileMetadata {
  constructor({ ...props }: FileMetadataProps) {
    super({
      ...props,
      type: new FileType(FileTypeValue.Pdf),
    });
  }

  foo() {
    // Doesn't exist on hydrated object.
  }
}
```

**Reply by mjroeleveld** [reply]
So, it does seem to call the constructor of `FileMetadata`, rather than `PdfFileMetadata`, even though the type is `Pdf`.

**Reply by B4nan** [reply]
I don't fully understand why you would want this, but I guess it should work regardless of that :] 

Could you provide a complete repro so I can try it myself? Including the entity where you use this, and some code that shows what you are doing with those entities, not just how you define them.

A few examples of good reproductions are here:

https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues

**Reply by B4nan** [reply]
I can see the problem will be that the discriminator value is mapped to your custom class, and the internals expects there to find some existing key from the discriminator map. So while I can see this as a rather easy fix, I still feel like you are causing more troubles to yourself with this approach for no reason. The polymorphic embeddables are exactly about distinguishing things by type, yet you want to have this (rather internal) property distinct on type level as well.

**Reply by mjroeleveld** [reply]
The reason why I do it this way is so that I can do validation in the `FileType` class, to validate input at runtime. That is not so clear in this example, but there are use cases where this is useful.

**Reply by mjroeleveld** [reply]
It's the [value object pattern](https://martinfowler.com/bliki/ValueObject.html). In this case the file type is a value object.

**Reply by mjroeleveld** [reply]
I removed the custom type, however the problem persists. See my last comment.

**Comment by mjroeleveld** [comment]
Repro:

```typescript
import {
  Embeddable,
  Embedded,
  Entity,
  Enum,
  MikroORM,
  PrimaryKey,
} from '@mikro-orm/core';
import { SqliteDriver } from '@mikro-orm/sqlite';

enum FileType {
  Image = 'image',
  Pdf = 'pdf',
}

@Embeddable({
  abstract: true,
  discriminatorColumn: 'type',
})
abstract class FileMetadata {
  @Enum()
  public readonly type: FileType;

  constructor({ type }: { type: FileType }) {
    this.type = type;
  }
}

@Embeddable({
  discriminatorValue: FileType.Pdf,
})
class PdfFileMetadata extends FileMetadata {
  constructor() {
    super({
      type: FileType.Pdf,
    });
  }

  foo() {
    // Doesn't exist on hydrated object.
  }
}

@Entity()
class File {
  @PrimaryKey()
  public readonly id: string;

  @Embedded()
  public readonly metadata: FileMetadata;

  constructor({ id, metadata }: { id: string; metadata: FileMetadata }) {
    this.id = id;
    this.metadata = metadata;
  }
}

describe('test', () => {
  let orm: MikroORM<SqliteDriver>;

  beforeAll(async () => {
    orm = await MikroORM.init({
      driver: SqliteDriver,
      dbName: ':memory:',
      entities: [File],
    });
    await orm.schema.createSchema();
  });

  afterAll(async () => {
    await orm.close(true);
  });

  test(`test`, async () => {
    const metadata = new PdfFileMetadata();
    const file = new File({
      id: '123',
      metadata,
    });
    await orm.em.persistAndFlush(file);

    const foundFile = await orm.em.findOne(File, '123');
    expect(foundFile.metadata.foo).toBeDefined();
  });
});
```

**Comment by mjroeleveld** [comment]
@B4nan the problem is the same for enums without custom types; it does not instantiate the right object regardless of the enum type. I.e., embeddables with `forceEntityConstructor: true` does not work. See my simplified repro above, I adjusted it for this bigger scope.
